src/domain_decomposition.hpp

```diff
@@ -113,42 +121,71 @@ class domain_decomposition {
 
     /// Returns the number of cells on the local domain.
     cell_size_type num_local_cells() const {
-        return cell_end()-cell_begin();
+        return gid_part_[domain_id_+1] - gid_part_[domain_id_];
     }
 
     /// Returns the number of cell groups on the local domain.
     cell_size_type num_local_groups() const {
-        return group_kinds_.size();
+        return groups_.size();
     }
 
     /// Returns meta data for a local cell group.
-    group_range_type get_group(cell_size_type i) const {
-        return {group_starts_[i], group_starts_[i+1], group_kinds_[i]};
+    const group_description& get_group(cell_size_type i) const {
+        EXPECTS(i<num_local_groups());
+        return groups_[i];
     }
 
     /// Tests whether a gid is on the local domain.
     bool is_local_gid(cell_gid_type i) const {
-        return i>=cell_begin_ && i<cell_end_;
+        return i>=gid_part_[domain_id_] && i<gid_part_[domain_id_+1];
     }
 
-    /// Return a partition of the cell gid over local cell groups.
-    gid_partition_type gid_group_partition() const {
-        return util::partition_view(group_starts_);
+private:
+    int num_domains_;
+    int domain_id_;
+    hw::node node_;
+    cell_size_type num_global_cells_;
+    std::vector<cell_gid_type> gid_part_;
+    std::vector<cell_kind> group_kinds_;
+    std::vector<group_description> groups_;
+};
+
+
+// The model needs to perform efficient lookup of properties associated
+// with cell gid. Currently the only property is the cell group of the
+// gid. The properties are stored in a gid_property struct.
+class gid_prop_map {
+public:
+    struct prop {
+        prop(cell_gid_type gid, cell_gid_type group): gid(gid), local_group(group) {}
+        cell_gid_type gid;
+        cell_gid_type local_group;
+        friend bool operator<(prop l, prop r) {return l.gid<r.gid;};
+    };
+
```

Comment:
You are right, what was I thinking.
This type has been removed completely.
* I used a hash table in the model to map gid to local cell group index, which I have marked with a big TODO to remove when the sampling interface is revamped.
* I use a flat vector in the `communicator` constructor, as suggested in your comment on that part of the code suggested.